src/puppetlabs/trapperkeeper/services/status/status_core.clj

```diff
@@ -127,6 +139,47 @@
   (let [status-map (service-status-map svc-version status-version status-fn)]
     (swap! status-fns-atom update-in [svc-name] conj status-map)))
 
+(defmacro with-timeout [timeout-s default & body]
+  `(let [f# (future (do ~@body))
+         result# (deref f# (* 1000 ~timeout-s) ~default)]
+     (future-cancel f#)
+     result#))
+
+(defn- maybe-explain
+  "Given the result of a call to schema.core/check, potentially unwrap it with
+  validation-error-explain if it is a ValidationError object. Otherwise, pass
+  the argument through."
+  [check-result]
+  (if (instance? schema.utils.ValidationError check-result)
+    (validation-error-explain check-result)
+    check-result))
+
+(schema/defn ^:always-validate guarded-status-fn-call
+  "Given a status check function, a status detail level, and a timeout in
+  seconds, this function calls the status function and handles three types of
+  errors:
+
+  * Status check timed out
+  * Status check through an Exception
+  * Status check returned a form that doesn't match the StatusCallbackResponse schema
+
+  In each error case, :is-running is set to :unknown and :status is set to a
+  string describing the error."
+  [status-fn :- StatusFn
+   level :- ServiceStatusDetailLevel
+   timeout :- schema/Int] :- StatusCallbackResponse
+   (let [unknown-response (fn [status] {:is-running :unknown
+                                        :status status})
+         timeout-response (unknown-response (format "Status check timed out after %s seconds" timeout))]
+     (with-timeout timeout timeout-response
+       (try
+         (let [result (status-fn level)]
+           (if-let [check-result (schema/check StatusCallbackResponse result)]
+             (unknown-response (format "Status check malformed: %s" (maybe-explain check-result)))
```

Comment:
I think `check-result` is also a mechanical description that doesn't give the reader any more information. What about `schema-failures` or `schema-problems`?